### PR TITLE
Render a list of developer features

### PR DESCRIPTION
**Pull request by nightnei**
Resolves https://github.com/Automattic/dotcom-forge/issues/5233

## Proposed Changes
We are moving "I am developer" checkbox from "Profile" to "Developer features" page - https://github.com/Automattic/wp-calypso/pull/86558.
Additionally we need to show the list of features which are available for developers, to provide the info of the checkbox "I am developer".
With this PR we are adding the list of that features to the page "Developer features".

## Testing Instructions
0) Checkout to the branch locally or open live branch
1) Open `http://calypso.localhost:3000/me/developer`
2) Assert that you see features and styles looks according to the design huXuof0X5yU9bqYvy6BPBW-fi-1970_1870
3) Assert that mobile version also looks goo according to the design huXuof0X5yU9bqYvy6BPBW-fi-1970_1870
4) Assert that loading-placeholder for "I am dev" checkbox looks also good: <br />Screenshot 2024-01-23 at 12 43 09

## The UI result (primarily for translators-team, to understand what it looks like)
Screenshot 2024-01-23 at 12 48 55

### CODE REVIEW

**Comment by nightnei** [comment]
To speed things up I have created draft. But I am a bit confused - which structure option will we use (1/2/3)? usehuXuof0X5yU9bqYvy6BPBW-fi-1052%3A2781

Screenshot 2024-01-19 at 13 05 19

**Comment by wongasy** [comment]
@nightnei A heads up that I rebased this branch, added finalized copy, mark as ready for review, and tagged the PR with `[Status] string freeze` to get the strings in for translation sooner. Please feel free to continue with development.

**Comment by a8ci18n** [comment]
This Pull Request is now available for translation here: https://translate.wordpress.com/deliverables/10966482

Some locales (Hebrew, Japanese) have been temporarily machine-translated due to translator availability. All other translations are usually ready within a few days. Untranslated and machine-translated strings will be sent for translation next Monday and are expected to be completed by the following Friday.

Thank you @nightnei for including a screenshot in the description! This is really helpful for our translators.

File: client/me/developer/features/index.tsx
Comment (wongasy) [inline client/me/developer/features/index.tsx]:
```suggestion
							{ linkLearnMore && <a href={ linkLearnMore } target="_blank" rel="noopener noreferrer">
								{ translate( 'Learn more' ) }
							</a> }
```

The last tile doesn't have a learn more link. Let's also ensure we track clicks to these links:

```
recordTracksEvent( 'calypso_me_developer_learn_more', { 'feature': title } )
```

And clicks to the inline links within the description should also be tracked.

File: client/me/developer/features/use-features-list.tsx
Comment (wongasy) [inline client/me/developer/features/use-features-list.tsx]:
```suggestion
```
We don't have a learn more link for this tile.

**Review by wongasy — CHANGES_REQUESTED** [review]
A few quick feedbacks:

* We need to adjust the title / subheading on mobile:

Screen Shot 2024-01-23 at 3 27 09 PM

* The inline links within a description should also be in pink:

Screen Shot 2024-01-23 at 3 33 29 PM

* We need to implement click tracking.

File: client/me/developer/features/index.tsx
Comment (wongasy) [reply client/me/developer/features/index.tsx]: Ah, I guess the condition should have been applied to `<div className="developer-features-list__item-learn-more">` instead. But you get the idea :)

File: client/me/developer/features/style.scss
Comment (wongasy) [inline client/me/developer/features/style.scss]:
```suggestion
	@media ( max-width: $break-large ) {
```

I would suggest breaking later. This is what we see on mobile (landscape):

| Break at `break-medium` | Break at `break-large` |
| ------------------------- | ------------------------- |
|Screen Shot 2024-01-23 at 3 46 58 PM|Screen Shot 2024-01-23 at 3 46 01 PM|

I wouldn't worry about adjusting the breakpoint for the sticky toggle. The suggestion here is mainly to ensure the text is easy to read.

File: client/me/developer/features/index.tsx
Comment (Aurorum) [inline client/me/developer/features/index.tsx:18]: It feels a bit strange to have these as `<div>` and not `<p>` tags. That would also stop you from needing to add the extra styles.

File: client/me/developer/features/style.scss
Comment (Aurorum) [inline client/me/developer/features/style.scss]:
```suggestion
			color: var(--color-text-subtle);
```

File: client/me/developer/features/style.scss
Comment (Aurorum) [inline client/me/developer/features/style.scss]: This will look broken on some colour schemes.

```suggestion
				color: var(--color-accent);
```

File: client/me/developer/features/use-features-list.tsx
Comment (Aurorum) [inline client/me/developer/features/use-features-list.tsx:20]: I suspect this might be temporary, but these would need to be the full URL!

File: client/me/developer/features/index.tsx
Comment (Aurorum) [inline client/me/developer/features/index.tsx]: With so many links here, I wonder if it might be nice to use `InlineSupportLink` - that might also add some consistency with other support links in Calypso?

File: client/me/developer/features/use-features-list.tsx
Comment (Aurorum) [inline client/me/developer/features/use-features-list.tsx:67]: Calypso tends to leave a space after dashes - eg.

https://github.com/Automattic/wp-calypso/blob/7749a6abe997eea80982902e8faff747c900c114/client/my-sites/earn/home.tsx#L129

https://github.com/Automattic/wp-calypso/blob/7749a6abe997eea80982902e8faff747c900c114/client/me/concierge/book/info-step.js#L175

I think it's worth being consistent with that.

```suggestion
				"Whenever you're stuck, whatever you're trying to make happen — our Happiness Engineers have the answers.",
```

File: client/me/developer/features/use-features-list.tsx
Comment (Aurorum) [inline client/me/developer/features/use-features-list.tsx:80]: Same as above.

```suggestion
				'Secure and maintain your site effortlessly with {{backupsLink}}real-time backups{{/backupsLink}}, advanced {{malwareScanningLink}}malware scanning and removal{{/malwareScanningLink}}, and continuous {{siteMonitoringLink}}site monitoring{{/siteMonitoringLink}} — ensuring peak performance and security at all times.',
```

File: client/me/developer/style.scss
Comment (Aurorum) [inline client/me/developer/style.scss]: Are these overrides necessary?

File: client/me/developer/style.scss
Comment (Aurorum) [inline client/me/developer/style.scss]: Is the colour override necessary?

File: client/me/developer/features/style.scss
Comment (Aurorum) [inline client/me/developer/features/style.scss:33]: If this is to keep things together, as I suspect, might wrapping the text in `preventWidows` be a better alternative in case the length of strings is different for other languages?

File: client/me/developer/features/style.scss
Comment (Aurorum) [inline client/me/developer/features/style.scss]: I'd personally recommend using the Card component rather than needing to re-do these styles here.

**Review by Aurorum — COMMENTED** [review]
Hey @nightnei 👋 I hope you don't mind the drive-by review, but I thought this section looked really cool and gave it a try. 

Not sure how complete this PR is, but I had a couple thoughts that might be of interest (feel free to ignore!). :)

File: client/me/developer/features/index.tsx
Comment (nightnei) [reply client/me/developer/features/index.tsx]: Removed the last "Learn more" link.
Added Track events.
`But you get the idea` - yep :) 

But I have one concern regarding tracking event: we have `title` property, but we can't use it as argument for the `recordTracksEvent`, since title is instance of `translate()` call, so we need to have plain `string`. I tried to create constants to avoid duplication of title, but variable are forbidden in `translate()`. So at the end of the day I duplicated the title. WDYT? Are you good with it?

https://github.com/Automattic/wp-calypso/pull/86648/commits/6b06317b6f3271944e608a6c7602e9511135d177#diff-a6d616c9253d9c26dd67a309b87c286b8596bea622c5de3b00a641991f152a0aR10-R11

File: client/me/developer/features/use-features-list.tsx
Comment (nightnei) [reply client/me/developer/features/use-features-list.tsx]: Removed

File: client/me/developer/features/style.scss
Comment (nightnei) [reply client/me/developer/features/style.scss]: Changed

File: client/me/developer/features/use-features-list.tsx
Comment (wongasy) [reply client/me/developer/features/use-features-list.tsx:20]: `localizeUrl()` turns the link to a full URL :)

File: client/me/developer/features/index.tsx
Comment (nightnei) [reply client/me/developer/features/index.tsx]: Regarding "title / subheading on mobile" - it turned out that it's ~bug~ feature 😅 https://github.com/Automattic/wp-calypso/blob/trunk/client/components/formatted-header/index.jsx#L47

As a quick fix - I added `break-word`.

File: client/me/developer/features/index.tsx
Comment (nightnei) [reply client/me/developer/features/index.tsx:18]: My typical approach for such casses: using `div`, since typically after some time somebody from marketing team asks to have a few paragraphs and in such case it's easy to add a few `<p>` instead existed `<div>`, otherwise it's necessary to change HTML. So I would keep it as is :)

File: client/me/developer/features/style.scss
Comment (nightnei) [reply client/me/developer/features/style.scss:31]: Could you please explain your vision regarding it?

File: client/me/developer/features/use-features-list.tsx
Comment (nightnei) [reply client/me/developer/features/use-features-list.tsx:20]: @Aurorum jsut to confirm - are we missing something or you missed `localizeUrl `?

File: client/me/developer/features/index.tsx
Comment (nightnei) [reply client/me/developer/features/index.tsx]: `InlineSupportLink ` doesn't have `onClick` argument, and we need to track event

File: client/me/developer/features/use-features-list.tsx
Comment (nightnei) [reply client/me/developer/features/use-features-list.tsx:67]: It was done intentionally, I also didn't know, but it turned out that avoiding spaces is more right way. But according to Google - there is no right way, both options are used by people :)

File: client/me/developer/features/style.scss
Comment (nightnei) [reply client/me/developer/features/style.scss:33]: No, it was added since designer used it in the design :)

File: client/me/developer/features/style.scss
Comment (nightnei) [reply client/me/developer/features/style.scss]: Make sense, thanks

**Comment by nightnei** [comment]
> Hey @nightnei 👋 I hope you don't mind the drive-by review, but I thought this section looked really cool and gave it a try.
> 
> Not sure how complete this PR is, but I had a couple thoughts that might be of interest (feel free to ignore!). :)

Thanks, I appreciate your review :)

**Comment by nightnei** [comment]
> Hey @nightnei 👋 I hope you don't mind the drive-by review, but I thought this section looked really cool and gave it a try.
> 
> Not sure how complete this PR is, but I had a couple thoughts that might be of interest (feel free to ignore!). :)

@Aurorum Thanks, I appreciate your review :)

File: client/me/developer/features/use-features-list.tsx
Comment (wongasy) [reply client/me/developer/features/use-features-list.tsx:67]: Here's an example we have no spaces around the em-dash: https://github.com/Automattic/wp-calypso/blob/fabdd100c4406dfd8dd350b473165c82142ecb8a/client/my-sites/domains/domain-management/list/google-domain-owner-banner.tsx#L19-L21

It would be nice to be consistent in all of Calypso, but by changing the copies now, the updated phrases will need to be re-translated again.

File: client/me/developer/features/style.scss
Comment (Aurorum) [reply client/me/developer/features/style.scss]: They're effectively the same colour, but all of Calypso tends to use that variable in cases like this and I don't think this section should be an exception. The advantage is consistency across colour schemes and that it can be easily updated in future (eg. this has happened in the past when changing the colour contrast ratio for a11y purposes).

File: client/me/developer/features/use-features-list.tsx
Comment (Aurorum) [reply client/me/developer/features/use-features-list.tsx:20]: I might be missing something, but this doesn't work for me!

ScreenRecording2024-01-24at13 39 53-ezgif com-video-to-gif-converter

File: client/me/developer/features/style.scss
Comment (Aurorum) [inline client/me/developer/features/style.scss]: I still think this needs updating to work on other colour schemes. 

```suggestion
			color: var(--color-accent);
```

File: client/me/developer/features/style.scss
Comment (Aurorum) [inline client/me/developer/features/style.scss]: Might this work? 

```suggestion
	font-family: $font-sf-pro-text;
```

File: client/me/developer/style.scss
Comment (Aurorum) [inline client/me/developer/style.scss:13]: This can cause problems on mobile.

Screenshot 2024-01-24 at 14 01 59

File: client/me/developer/features/use-features-list.tsx
Comment (nightnei) [reply client/me/developer/features/use-features-list.tsx:20]: Thanks for explanation. It doesn't work locally, but if you replace `localhost` with `wordpress.com` it will be ok. So it production it will work.

File: client/me/developer/style.scss
Comment (Aurorum) [inline client/me/developer/style.scss]: It'd be nice to use a colour variable, not least because this looks very strange on some colour schemes where there is no blue. I'd recommend this:

```suggestion
	background: var(--color-accent-0);
```

Screenshot 2024-01-24 at 14 06 12

File: client/me/developer/features/use-features-list.tsx
Comment (Aurorum) [reply client/me/developer/features/use-features-list.tsx:20]: Ah, I see, thanks for explaining! Will it still work for folks on https://horizon.wordpress.com/ or https://wpcalypso.wordpress.com/ though?

Perhaps not necessary, but I feel a feature like this might be disproportionately used by folks who use Horizon or WPCalypso.

File: client/me/developer/features/style.scss
Comment (nightnei) [reply client/me/developer/features/style.scss]: I have already updated it

File: client/me/developer/features/style.scss
Comment (nightnei) [reply client/me/developer/features/style.scss]: There is a lot of options of using `font-family` in Calypso, but yeah, I agree, this option I like more, thanks

File: client/me/developer/style.scss
Comment (nightnei) [reply client/me/developer/style.scss]: I will think about it, but it looks totally different on default theme, and the designer used exactly the current color in the design

Comment (Aurorum) [reply client/me/developer/style.scss]: Ah, oops, you're right! It's not really for this purpose, but this might work:

```suggestion
	background: var(--color-link-0);
```

Assuming Design agrees, I do think a colour variable is better here.

File: client/me/developer/features/style.scss
Comment (Aurorum) [inline client/me/developer/features/style.scss:9]: Perhaps intentional, but do you maybe want to override this padding when the screen breaks on mobile? Otherwise there's a bit much unnecessary space at the bottom of the screen.

File: client/me/developer/features/index.tsx
Comment (Aurorum) [inline client/me/developer/features/index.tsx:26]: This is a design choice, but do you want to use `ExternalLink` here? That would indicate that the link opens in another tab, while also adding some screen reader text which is important for accessibility.

File: client/me/developer/style.scss
Comment (Aurorum) [reply client/me/developer/style.scss:13]: Still a problem - might be worth reducing the font size on mobile?

**Review by Aurorum — APPROVED** [review]
All my feedback has been addressed - this is looking really good! I love the placeholder toggle; obviously not for this PR, but would be interesting to see if something similar could maybe be added to the component as a loading state one day since it looks great. 

Couple of non-blocking comments - these are my personal preferences:

1. I'd prefer if the support links were the full link - this feels like the sort of feature which would be disproportionately used on localhost/WPCalypso/Horizon and there's no real harm in doing so. :) I feel it's bad practice to rely on `localizeUrl` like this.  

2. I still feel very strongly that the colour at the bottom should be a variable that changes with colour schemes. Blue looks odd when there is no other blue in the scheme (obvs not my call though!). I'd recommend `var(--color-link-0)`.

**Edit:** just saw you added the variable - thank you!

File: client/me/developer/features/handle-click-link.tsx
Comment (wongasy) [inline client/me/developer/features/handle-click-link.tsx]:
```suggestion
	const matches = /\/([^\/]+)$/.exec(link);
	const pageSlug = matches[1] ?? event.currentTarget.href;

    recordTracksEvent( 'calypso_me_developer_learn_more', { feature: pageSlug } )
```

The current implementation expects a support doc link at all times, I'd like to propose a more flexible approach where we grab the last segment of the URL. Chances are, we'll likely be linking out to a support doc, so I guess what you have here is likely OK. But thought I'd make the suggestion regardless.

I'll leave it up to you to decide.

File: client/me/developer/features/index.tsx
Comment (nightnei) [reply client/me/developer/features/index.tsx:26]: I think it can confuse a bit, since the description of the component is `Link to an external resource.`, but it's not an external resource.

**Comment by nightnei** [comment]
@Aurorum `Still a problem - might be worth reducing the font size on mobile?`
Hm, It would be a problem at least for iPhone 5S, but Apple doesn't support, so I would keep it as is.
Also, according to the Calypso break points, for mobile we consider 480px, so I suppose that it's the minimum what is should be supported.

File: client/me/developer/features/handle-click-link.tsx
Comment (nightnei) [reply client/me/developer/features/handle-click-link.tsx]: Such approach works a bit anotehr way - it take the last item in path, for example for `/support/domains/https-ssl`.

And also, I suppose that in the future it's possible to have some case e.g. 
1) `/support/domains/lorem-ipsum`
2) `/support/emails/lorem-ipsum`

I think it's better to go the current way, with providing `/domains/https-ssl`, instead of `https-ssl`

WDYT?

Comment (wongasy) [reply client/me/developer/features/handle-click-link.tsx:12]: I see the case you're covering here. Let's stick with your approach, I would just ask that we remove the leading `/` (e.g. report `site-monitoring` instead of `/site-monitoring`).

Comment (wongasy) [reply client/me/developer/features/handle-click-link.tsx]: I see the case you're covering here. Let's stick with your approach, I would just ask that we remove the leading `/` (e.g. report `site-monitoring` instead of `/site-monitoring`).

I would also rename the variable `title` to `featureSlug` so it better describes what the variable represents.

File: client/me/developer/style.scss
Comment (wongasy) [reply client/me/developer/style.scss:13]: Looks like @nightnei has adjusted the font size and word break is no longer an issue on mobile.

Screen Shot 2024-01-24 at 11 26 53 AM

File: client/me/developer/features/index.tsx
Comment (wongasy) [inline client/me/developer/features/index.tsx]:
```suggestion
					<Card className="developer-features-list__item" key={ urlToSlug( linkLearnMore ) }>
```

We need to supply a `key` property to the `<Card>` to resolve this warning:

Screen Shot 2024-01-24 at 11 39 42 AM

We shouldn't be using an index as a node key (see https://github.com/Automattic/wp-calypso/pull/86186#discussion_r1447053718), so I'm proposing that we generate a slug using the learn more link and use that as the key.

**Review by wongasy — APPROVED** [review]
LGTM! This is good to go pending the final comments in this round of review.

* The page renders as expected for desktop:

Screen Shot 2024-01-24 at 11 27 05 AM

* The page renders as expected for mobile:

Screen Shot 2024-01-24 at 11 26 53 AM

* Tracking events are also being fired (as noted in the comment, let's remove the leading `/` from the `feature` prop):

Screen Shot 2024-01-24 at 11 27 26 AM

File: client/me/developer/features/index.tsx
Comment (nightnei) [reply client/me/developer/features/index.tsx]: It's not recommended way, but when there is no better option then it's ok to use index as a key, ofc, if you are sure that it's safe to use it in some specific case. For example in our case:
1) We don't have linkLearnMore for the last item, so we can't use it
2) It's totally safe to use index in this case
3) Index in this case is really the best approach, since we don't use extra function (e.g. urlToSlug) and it's literalyt tiny-tiny-tiny optimization 😅

Comment (wongasy) [reply client/me/developer/features/index.tsx]: @nightnei I was recommended by @tyxla to stay away from using indices as keys. I'll defer to you and Marin to decide on the best approach to move forward with this.

**Comment by Aurorum** [comment]
@nightnei 

> Also, according to the Calypso break points, for mobile we consider 480px, so I suppose that it's the minimum what is should be supported.

Calypso does style a lot of things below 480px, plus a lot of devices can be smaller. But looks like this has been fixed now! :)

Am curious for your thoughts on not relying on `localizeUrl` though - I still feel a little strongly on that since it’d break locally, on Horizon and on WPCalypso.

**Comment by wongasy** [comment]
@nightnei Confirmed that the leading `/` is removed from the tracking details:

Screen Shot 2024-01-24 at 3 06 06 PM

**Review by wongasy — APPROVED** [review]

File: client/me/developer/features/index.tsx
Comment (nightnei) [reply client/me/developer/features/index.tsx]: Ok, if I am missing something - I would appreciate you pointing me out to it. But I am 99.999% positive 🙂 that the index, in this case, is the best and right approach. In two words - we are working with static, hardcoded array at the frontend, so some complex `key` here is literally would be overkill. More details are described above.

Comment (tyxla) [reply client/me/developer/features/index.tsx]: @nightnei, @wongasy is actually right. We should never use indexes as keys, and here's why: https://react.dev/learn/rendering-lists#why-does-react-need-keys

I understand that this is a hardcoded list right now, but what if it gets dynamic at some point? How will folks know that they should update the key to not use indexes?

Also, if you're working with a hardcoded array, then you can generate a key out of the existing array, or if necessary, even add an additional property to each array item.

Comment (nightnei) [reply client/me/developer/features/index.tsx]: Thanks for explanation of your thoughts.

>I understand that this is a hardcoded list right now, but what if it gets dynamic at some point?
If it's dynamic, then maybe we will receive also something like `id` from backend, or maybe we will receive already translated `title` and will use it.

From new documentations:
`You might be tempted to use an item’s index in the array as its key. In fact, that’s what React will use if you don’t specify a key at all.`

From old documentation (https://legacy.reactjs.org/docs/lists-and-keys.html):
`When you don’t have stable IDs for rendered items, you may use the item index as a key as a last resort`


`or if necessary, even add an additional property to each array item.` - I see it only as 1 available option in this case, and I still think that in this case `index` is better approach, but I don't mind going this way. Should I do it? Or do you have another approach?

Comment (tyxla) [reply client/me/developer/features/index.tsx]: > You might be tempted to use an item’s index in the array as its key. In fact, that’s what React will use if you don’t specify a key at all.

And you'll get warnings 😉 

> When you don’t have stable IDs for rendered items, you may use the item index as a key as a last resort

I agree, but I don't think this is the last resort - you definitely have other options since you have full control over the array. 

> `or if necessary, even add an additional property to each array item.` - I see it only as 1 available option in this case, and I still think that in this case index is better approach, but I don't mind going this way. Should I do it? Or do you have another approach?

I'd suggest adding an `id` property to the array, and using that. I can definitely see this list becoming dynamic at some point, and us missing to update the `key` to something better than the index.

Comment (nightnei) [reply client/me/developer/features/index.tsx]: Ok, let's go this way, I added `id` properties to the array.
Thanks @tyxla for the response and your time :)

**Review by danielbachhuber — COMMENTED** [review]
How did the toggle end up at the bottom of the page? Given it's the most important element of the page, can we put it under the heading?

**Comment by a8ci18n** [comment]
Translation for this Pull Request [has now been finished](https://translate.wordpress.com/deliverables/overview/10966482).

**Comment by wongasy** [comment]
>How did the toggle end up at the bottom of the page? Given it's the most important element of the page, can we put it under the heading?

@danielbachhuber It was moved to the bottom as recommended by design for visual balance: 5233-gh-Automattic/dotcom-forge#issuecomment-1903995631 

Although, I do agree it makes sense to ensure that's the first thing users see on the page. @nightnei Let's move the toggle to the top of the page as per @javierarce's previous iteration of the design 5233-gh-Automattic/dotcom-forge#issuecomment-1903536666

**Comment by javierarce** [comment]
> Although, I do agree it makes sense to ensure that's the first thing users see on the page. @nightnei Let's move the toggle to the top of the page as per @javierarce's previous iteration of the design 5233-gh-Automattic/dotcom-forge#issuecomment-1903536666

Before we make this change, let's get a consensus from more people on the design team. I think the toggle at the bottom, highlighted with a light blue background, addresses the main concern of visibility and looks better than the version with the toggle at the top. But I'll ask around to get feedback from the team.

**Comment by javierarce** [comment]
I still think the toggle at the bottom, highlighted with a light blue background, addresses the main concern of visibility and looks better and more balanced than the version with the toggle at the top, but I don't want to block the release.

**Comment by danielbachhuber** [comment]
What happens if the user has their browser set to only 600px tall?

**Comment by wongasy** [comment]
I do see that visually, it looks better at the bottom, but I also see Daniel's point of prioritizing the toggle to ensure it's in view. Let's move forward with moving the toggle to the top, and we can iterate later if needed!

**Pull request merged**
